src/API/Middlewares/RequestValidationMiddleware.ts

```diff
@@ -0,0 +1,27 @@
+import { IllegalArgumentException } from '@po.et/poet-js'
+import * as Joi from 'joi'
+interface RequestValidationParams {
+  readonly query?: (values: object) => {}
+  readonly body?: (values: object) => {}
+  readonly params?: (values: object) => {}
+  readonly options?: object
+}
+
+export const RequestValidationMiddleware = (requestValidationParams: RequestValidationParams) => async (
+  context: any,
+  next: () => Promise<any>
+) => {
+  try {
+    const { query, body, params, options } = requestValidationParams
+    /* tslint:disable:no-unused-expression */
+    body && (await Joi.validate(context.request.body, body(context.request.body), options))
+    query && (await Joi.validate(context.query, query(context.query), options))
+    params && (await Joi.validate(context.params, params(context.params), options))
+    return next()
+  } catch (exception) {
+    if (context.logger)
+      context.logger.error({ exception }, 'Error Caught at Middleware - Will Return Internal Server Error')
```

Comment:
Does `Joi.validate` throw if the validation fails?

If so, this isn't an error and we shouldn't log it as such. Don't know whether we care about logging it at all, even. Maybe log it as `debug`?

If we decide to keep this log, the logging message should be updated too. It's not an Internal Server Error, and it's not an Error. I'm thinking `RequestValidationMiddleware rejected a request.`

Reply:
Does Joi.validate throw if the validation fails?

yes, `Joi.validate ` throw an error

What do you think, it's important to log it?

Reply:
I'd go with leaving it there out of caution, but just under `debug` so we don't get a flood of error messages that trigger alarms.

Reply:
ok